basics/create-a-solver/create-a-solver.jmd

```diff
@@ -0,0 +1,205 @@
+the problem, as it contains details regarding its description (i.e function to
+be minimized, starting point, number of variables, etc.), while a
+**GenericExecutionStats** is responsible for carrying information over how the
+problem was solved (i.e number of iterations, time elapsed, number of function
+evaluations, etc.).
+
+Another important thing to highlight, is that we'll be using the **ADNLPModel**
+struct to run our solver. By default, the package responsible for calculating
+derivatives is [ForwardDiff](https://github.com/JuliaDiff/ForwardDiff.jl) since
+that's the built in method for differentation in this **AbstractNLPModel**
+implementation.
+
+When creating a new JSO compatible solver, it's also important to manipulate
+the output data to match that of **GenericExecutionStats**. The ``status``
+attribute tells us about how our solver did given our input problem. It is
+possible that the number of iterations has exceed the max limit, or the time it
+took for it to solve was greater than allowed. Either way, the ``status`` is an
+important attribute when benchmarking and testing, as it allows the user to keep
+track of how a solver behaves over a specific problem.
+
+If we want to run our solver, we simply need to create a optimization problem
+```julia
+  f(x) = (x[1]^2 + x[2]^2)^2
+  problem = ADNLPModel(f, [1.0,2.0])
+  output = newton(problem)
+```
+whose information regarding its solution is stored in ``output``.
+```julia
+  print(output)
+```
+With a solver in hands, we can start to do more interesting things, such as
+benchmarking and comparing our ```newton``` method to other solvers. Since JSO
+contains a lot of useful packages, we'll be using a collection of those to make
+our comparison easier. The most important one by far is [SolverBenchmarks](https://github.com/JuliaSmoothOptimizers/SolverBenchmark.jl), a
+benchmarking package designed for manipulating DataFrames that contain statistics
+regarding the execution of a specific sover given a set of problems.
+
+Let us compare our ```newton``` method to a solver contained in the
+[JSOSolvers](https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl) package, namely the ```lbfgs``` one.
+Firstly, we'll need to gather our data regarding each solver.
+```julia
+using SolverBenchmark, JSOSolvers
+solvers = Dict(:newton=>newton, :lbfgs=>lbfgs)
+f1(x) = x[1]^2 + x[2]^2
+f2(x) = (1 - x[1])^2 + 100(x[2] - x[1]^2)^2
+f3(x) = (x[1]^2 + x[2] - 11) + (x[1] + x[2]^2 - 7)^2
+f4(x) = 0.26(x[1]^2 + x[2]^2) - 0.48*x[1]*x[2]
+test_functions = [f1, f2, f3, f4]
+problems = (ADNLPModel(i, 2*ones(2),name="Problem $i") for i in test_functions)
+stats = bmark_solvers(solvers, problems)
+```
+The ``bmark_solvers`` function is also part of the **SolverTools** package, and
+runs our solver on a set of problems, in our case ``f1``, ``f2``, ``f3`` and ``f4``.
+Lastly, it returns a DataFrame of statistics that we can manipulate using
```

Comment:
It returns a dictionary of `solver => DataFrame`, where the DataFrame is the execution data of that solver in the problems.